qiskit/circuit/quantumcircuit.py

```diff
@@ -4319,108 +4296,28 @@ def assign_parameters(  # pylint: disable=missing-raises-doc
         if inplace:
             target = self
         else:
+            if not isinstance(parameters, dict):
+                # We're going to need to access the sorted order wihin the inner Rust method on
+                # `target`, so warm up our own cache first so that subsequent calls to
+                # `assign_parameters` on `self` benefit as well.
+                _ = self._data.parameters
             target = self.copy()
             target._increment_instances()
             target._name_update()
 
-        # Normalize the inputs into simple abstract interfaces, so we've dispatched the "iteration"
-        # logic in one place at the start of the function.  This lets us do things like calculate
-        # and cache expensive properties for (e.g.) the sequence format only if they're used; for
-        # many large, close-to-hardware circuits, we won't need the extra handling for
-        # `global_phase` or recursive definition binding.
-        #
-        # During normalisation, be sure to reference 'parameters' and related things from 'self' not
-        # 'target' so we can take advantage of any caching we might be doing.
-        if isinstance(parameters, dict):
+        if isinstance(parameters, collections.abc.Mapping):
```

Comment:
Are there cases people are passing dict likes to `assign_parameters`? Also I assume you're not worried about the extra runtime overhead of the registration hooks of the collections types here.

Reply:
No, and in fact it would have been impossible before because of the type hint.  I'd loosened this because a) that matches the documentation and b) I think I was passing the `ParameterBindsDict` custom object recursively into subcalls at some point (though I think I changed the logic subsequently).